src/cargo-kani/src/call_cbmc.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use anyhow::{bail, Result};
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use crate::session::KaniSession;
+
+impl KaniSession {
+    /// Verify a goto binary that's been prepared with goto-instrument
+    pub fn run_cbmc(&self, file: &Path) -> Result<PathBuf> {
+        let output_filename = crate::util::append_path(file, "cbmc_output");
+
+        {
+            let mut temps = self.temporaries.borrow_mut();
+            temps.push(output_filename.clone());
+        }
+
+        let args: Vec<OsString> = self.cbmc_flags(file)?;
+
+        // TODO get cbmc path from self
+        let mut cmd = Command::new("cbmc");
+        cmd.args(args);
+
+        if self.args.output_format == crate::args::OutputFormat::Old {
+            let result = self.run_terminal(cmd);
+            if !self.args.allow_cbmc_verification_failure {
+                result?;
+            }
+        } else {
+            // extra argument
+            cmd.arg("--json-ui");
+            let result = self.run_redirect(cmd, &output_filename)?;
+
+            // regardless of success or failure, first we need to print:
+            self.format_cbmc_output(&output_filename)?;
+
+            if !result.success() && !self.args.allow_cbmc_verification_failure {
+                bail!("cbmc exited with status {}", result);
+            }
+        }
+
+        Ok(output_filename)
+    }
+
+    /// used by call_cbmc_viewer, invokes different variants of CBMC.
+    // TODO: this could use some cleanup and refactoring.
+    pub fn call_cbmc(&self, args: Vec<OsString>, output: &Path) -> Result<()> {
+        // TODO get cbmc path from self
+        let mut cmd = Command::new("cbmc");
+        cmd.args(args);
+
+        let result = self.run_redirect(cmd, output)?;
+
+        if !result.success() {
+            bail!("cbmc exited with status {}", result);
+        }
+        // TODO: We 'bail' above, but then ignore it in 'call_cbmc_viewer' ...
+
+        Ok(())
+    }
+
+    /// "Internal," but also used by call_cbmc_viewer
```

Comment:
I suppose everything could be `pub(crate)` instead of `pub` really. Is there an advantage to doing so?

Reply:
It was more like a question whether you wanted to make it explicit here given the internal comment.

I'm not sure if it makes any difference when it comes to dead code analysis since this is only compiled into a binary.